test/scripts/add_newline.py

```diff
@@ -38,13 +38,23 @@ def add_newline(file: str, root_dir: str) -> bool:
 
         contents = current_file.read(1)
         if not contents == "\n":
-            current_file.write("\n")
+            if should_fix:
+                current_file.write("\n")
+            else:
+                print(f"{file} lacks ending newline")
             return True
```

Comment:
It took me quite a while to get that `True` means "there was no newline but it might have been added" and `False` means "there is a newline".

I think with the latest change the logic became quite convoluted, maybe you could make it more obvious by:
- make `add_newline` return `True` if there is a newline, or `should_fix` was given and the newline was successfully added
- return `False` if there is no newline and `should_fix` was not given

This way you can also simplify the check in line 59, because you error exit only if one `add_newline` returned false